lib/metasploit/framework/credential_collection.rb

```diff
@@ -1,233 +1,292 @@
-  #
-  # @yieldparam credential [Metasploit::Framework::Credential]
-  # @return [void]
-  def each
-    if pass_file.present?
-      pass_fd = File.open(pass_file, 'r:binary')
+    # @!attribute prepended_creds
+    #   List of credentials to be tried before any others
+    #
+    #   @see #prepend_cred
+    #   @return [Array<Credential>]
+    attr_accessor :prepended_creds
+
+    # @!attribute realm
+    #   @return [String]
+    attr_accessor :realm
+
+    # @option opts [Boolean] :blank_passwords See {#blank_passwords}
+    # @option opts [String] :pass_file See {#pass_file}
+    # @option opts [String] :password See {#password}
+    # @option opts [Array<Credential>] :prepended_creds ([]) See {#prepended_creds}
+    # @option opts [Boolean] :user_as_pass See {#user_as_pass}
+    # @option opts [String] :user_file See {#user_file}
+    # @option opts [String] :username See {#username}
+    # @option opts [String] :userpass_file See {#userpass_file}
+    def initialize(opts = {})
+      opts.each do |attribute, value|
+        public_send("#{attribute}=", value)
+      end
+      self.prepended_creds     ||= []
+      self.additional_privates ||= []
     end
 
-    prepended_creds.each { |c| yield c }
+    # Adds a string as an addition private credential
+    # to be combined in the collection.
+    #
+    # @param [String] private_str the string to use as a private
+    # @return [void]
+    def add_private(private_str='')
+      additional_privates << private_str
+    end
 
-    if username.present?
-      if password.present?
-        yield Metasploit::Framework::Credential.new(public: username, private: password, realm: realm, private_type: private_type(password))
+    # Add {Credential credentials} that will be yielded by {#each}
+    #
+    # @see prepended_creds
+    # @param cred [Credential]
+    # @return [self]
+    def prepend_cred(cred)
+      prepended_creds.unshift cred
+      self
+    end
+
+    # Combines all the provided credential sources into a stream of {Credential}
+    # objects, yielding them one at a time
+    #
+    # @yieldparam credential [Metasploit::Framework::Credential]
+    # @return [void]
+    def each
+      if pass_file.present?
+        pass_fd = File.open(pass_file, 'r:binary')
       end
-      if user_as_pass
-        yield Metasploit::Framework::Credential.new(public: username, private: username, realm: realm, private_type: :password)
+
+      prepended_creds.each { |c| yield c }
+
+      if password.present?
+        yield Metasploit::Framework::Credential.new(private: password, realm: realm, private_type: private_type(password))
       end
       if blank_passwords
-        yield Metasploit::Framework::Credential.new(public: username, private: "", realm: realm, private_type: :password)
+        yield Metasploit::Framework::Credential.new(private: "", realm: realm, private_type: :password)
       end
       if pass_fd
         pass_fd.each_line do |pass_from_file|
```

Comment:
No if `pass_file.present?` is `false` then yes `pass_fd` will never be set, but the check on `pass_fd` will eval to `false` because... Ruby. The variable will be `nil` causing this to be safe. In your test you don't have a conditional check that optionally sets `a`.

See:
```
2.6.6 :001 > if false
2.6.6 :002 >   a = "A"
2.6.6 :003 > end
 => nil 
2.6.6 :004 > if a
2.6.6 :005 >   puts a
2.6.6 :006 > end
 => nil 
2.6.6 :007 > 
```

Reply:
Thanks for pointing this out. Wasn't aware that Ruby would define this and set it to `nil` under these circumstances; appreciate you showing me this. Please ignore my comments on this seeing as they are incorrect.